upstream: outlier detection for non 5xx codes

cpakulski opened the pull request:
Commit Message: outlier detection for non 5xx codes
Additional Description: 
Added to a cluster HTTP a specific option to define an http matcher. When a response matches the defined matcher, the response is treated as error, regardless of the status code. The error is reported to outlier detection as 5xx code. If a response does not match the matcher it is treated as success and forwarded to outlier detection as code 200.
Risk Level: Low (no change unless matcher is defined)
Testing: Added unit and integration tests
Docs Changes: yes
Release Notes: yes
Platform Specific Features: n/a
Fixes #18789

wbpcode commented on api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto:
```diff
@@ -174,4 +176,16 @@ message HttpProtocolOptions {
   // [#not-implemented-hide:]
   // [#extension-category: envoy.http.header_validators]
   config.core.v3.TypedExtensionConfig header_validation_config = 7;
+
+  message HttpEvents {
+    // Matcher for response headers.
+    config.common.matcher.v3.MatchPredicate match = 1 [(validate.rules).message = {required: true}];
+
+    // Code which should be reported to the outlier detection if response headers matched the matcher.
+    envoy.type.v3.StatusCode map_to = 2 [(validate.rules).enum = {defined_only: true}];
+  }
+
+  // List of matchers for response headers along with codes to be reported to outlier detection.
+  // The first matcher which returns matching success is applied.
+  repeated HttpEvents outlier_detection = 8;
```
Why we need a list of `HttpEvents`? I think single match should enough to match an error?

```
message OutlierDetection {
  config.common.matcher.v3.MatchPredicate failure_matcher = 1 [(validate.rules).message = {required: true}];
}

OutlierDetection outlier_detection = 8;
```

wbpcode commented on envoy/upstream/upstream.h:
```diff
@@ -1013,6 +1013,9 @@ class ClusterInfo : public Http::FilterChainFactory {
    */
   virtual const envoy::config::cluster::v3::Cluster::CommonLbConfig& lbConfig() const PURE;
 
+  virtual absl::optional<uint64_t>
+  processHttpForOutlierDetection(Http::ResponseHeaderMap& reponse) const PURE;
```
```suggestion
  virtual absl::optional<bool>
  checkFailureResponse(Http::ResponseHeaderMap& reponse) const PURE;
```

return an `absl::optional<bool>` make more sense. It's unecessary to return http code.

cpakulski replied:
Agree.

wbpcode reviewed (commented):
Thanks for the update. I didn't get why we need a `map_to`. Why not return a bool by single `failure_matcher`. Which scenario is the single `failure_matcher` couldn't cover?

wbpcode commented on envoy/upstream/upstream.h:
```diff
@@ -1013,7 +1013,7 @@ class ClusterInfo : public Http::FilterChainFactory {
 
-  virtual absl::optional<uint64_t>
+  virtual absl::optional<bool>
   processHttpForOutlierDetection(Http::ResponseHeaderMap& reponse) const PURE;
```
Rather then adding such new method at ClusterInfo. I think it would be better to expose the `httpProtocolOptions()` here.

cpakulski replied:
That is probably a good idea. `ClusterInfo` should be just a config repository without methods with logic.

wbpcode commented on api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto:
```diff
@@ -181,8 +181,8 @@ message HttpProtocolOptions {
-    // Code which should be reported to the outlier detection if response headers matched the matcher.
-    envoy.type.v3.StatusCode map_to = 2 [(validate.rules).enum = {defined_only: true}];
+    // Determines if matching indicate error or non-error. Defaults to false (error).
+    bool success_on_match = 2;
```
I prefer the `xds.type.matcher.v3.Matcher` rather then `config.common.matcher.v3.MatchPredicate` because `xds.type.matcher.v3.Matcher` support CEL and is easier to support various logical operators.

cpakulski replied:
```
failure_matcher:
  or_match:
    rules:
    - http_response_headers_match:
        headers:
        - name: ":status"
          string_match:
            safe_regex:
              regex: "4.."
    - and_match:
        rules:
        - http_response_headers_match:
            headers:
            - name: ":status"
              string_match:
                safe_regex:
                  regex: "5.."
        - not_match:
            http_response_headers_match:
              headers:
              - name: ":status"
                string_match:
                  safe_regex:
                    regex: "502|503|504"
```

So let's see what happens when 502 is returned. When it is matched against that matcher, the result will be NOT_MATCHED. So 502 will be forwarded to outlier and consecutive counter will be incremented, and we do not want that as described in #38311. What we need is MATCH result but with some sort of a flag which says treat it as success when reporting to outlier. 
One option is attribute `success_on_match` or just two individual matchers:
```
xds.type.matcher.v3.Matcher failure_matcher = 3;
xds.type.matcher.v3.Matcher success_matcher = 4;
```
Then we do not need `success_on_match` attribute and we do not need `repeated` matchers.

cpakulski replied:
> I do don't want to enhance the old outlier, but if we do it, then not match means success and will be convert to 200? The only case where we keep the original status code is the case where the matcher is not configured?
> 

Actually no, NOT_MATCHED means that the original status code is reported to "old" outlier. MATCHED means that  either error or success will be reported to outlier, depending if match happened on `failure_matcher` or `success_matcher` (or depending on `success_on_match` flag if we decide to use boolean flag).

>
> So, if the the result will be NOT_MATCHED, I think the 502 will be treat as success because the matcher is failure_matcher? And then, it will be convert to 200?

It should be:
> So, if the the result will be **_MATCHED_**, I think the 502 will be treat as success because the matcher is _**success_matcher**_? And then, it will be convert to 200?

Yes, but only when reporting to outlier. The downstream client will receive original status code. We use here 500 for error and 200 for success only because this is how outlier understands what is an error/success. If it was using booleans, we would just report true/false.

cpakulski replied:
> it's not very hard to accept a failure_matcher and a success_matcher. But could you kindly explain more about why? In my mind, if single failure_matcher is configured, that means the failure_matcher will take over the determination of the failure/success. Or if the failure_matcher is not configured, the legacy behavior will be used.

I think it is possible to use only `failure_matcher` but it will be a bit more difficult to understand from users' point of view. I envisioned that `failure_matcher` will be used only to **add** additional errors to already existing ones (5xx), not completely redefine what codes are treated as errors. (And `success_matcher` would be used to **remove** status codes from default error codes). So, if a user wants to treat 4xx as errors, the user needs to define `failure_matcher` for 4xx only. If the `failure_matcher` decides what is error and what is not error, then the user must define matcher for 4xx and 5xx, otherwise 5xx will be treated as success.

So, in essence there are two approaches here:
- always use legacy behaviour and use `failure_matcher` and `success_matcher` to **add** or **remove** status codes from default set.
- use only `failure_matcher` to completely overwrite legacy behavior. If `failure_matcher` is not specified, legacy behaviour is takes place (only 5xx are treated as errors). If a `failure_matcher` is defined, only codes matching the matcher will be treated as error.

wbpcode replied:
> I think it is possible to use only failure_matcher but it will be a bit more difficult to understand from users' point of view. I envisioned that failure_matcher will be used only to add additional errors to already existing ones (5xx), not completely redefine what codes are treated as errors. (And success_matcher would be used to remove status codes from default error codes). So, if a user wants to treat 4xx as errors, the user needs to define failure_matcher for 4xx only. If the failure_matcher decides what is error and what is not error, then the user must define matcher for 4xx and 5xx, otherwise 5xx will be treated as success.

I think this is one of our differenet perspectives. IMO, the multiple matcher bring much more complexity because the users to aware the multiple matcher's result and the original codes to figure out the final success/failure result.

I will prefer the second approach in your list. Simplely derterming the result based on legacy codes or new matcher would be more intuitive in most cases.

cpakulski replied:
I will implement a single matcher and see how it goes with the rest of the logic. Thanks.

cpakulski replied:
I can convert the matcher to `xds.type.matcher.v3.Matcher` if required. I have no opinion which one is better or faster. Matching happens on each response, so performance is important.

wbpcode commented on api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto:
```diff
@@ -174,4 +175,8 @@ message HttpProtocolOptions {
   // [#extension-category: envoy.http.header_validators]
   config.core.v3.TypedExtensionConfig header_validation_config = 7;
+
+  // If specified, only responses matching the matcher will be treated by outlier detection as errors.
+  // If not specified, only 5xx codes are treated by outlier detection as errors.
+  config.common.matcher.v3.MatchPredicate outlier_detection_error_matcher = 8;
```
I actually prefer you previous design about add a new `OutlierDetection` message. We may could add new field at there in the future if we aware more feature requirements. (If I didn't present this clearly before, sorry! 😞 )

```
message OutlierDetection {
  config.common.matcher.v3.MatchPredicate failure_matcher = 1;
}
```

wbpcode reviewed (commented):
Only one minor comment to the API. Once we get agreement to the API, that should pretty quick to complete the review the code implemention. I think maybe we can land this PR before this weekend or next Tuesday. :)

cpakulski commented:
/retest

cpakulski commented:
/retest

cpakulski commented:
@wbpcode Thanks for approving the API. I added integration tests and updated docs. I believe that it is ready for another review (moving it out of draft).

cpakulski commented on source/extensions/common/matcher/matcher.h:
```diff
@@ -87,7 +87,7 @@ class Matcher {
    *                 size as the match tree vector (see above).
    */
-  virtual void onHttpRequestHeaders(const Http::RequestHeaderMap& request_headers,
+  virtual void onHttpRequestHeaders(const Envoy::Http::RequestHeaderMap& request_headers,
```
I run into the namespace conflict. I do not remember exactly where it was (but can dig it out if necessary), where another `Http` namespace was used and confused the compiler to search for `RequestHeaderMap` in that other namespace, not main Envoy one.

cpakulski commented on source/common/router/router.cc:
```diff
@@ -1656,16 +1656,38 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::ResponseHeaderMapPt
     }
   }
 
   maybeProcessOrcaLoadReport(*headers, upstream_request);
```
Sure. NP.

cpakulski commented on source/common/router/router.cc:
```diff
@@ -1656,16 +1656,38 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::ResponseHeaderMapPt
   // handle the case when an error grpc-status is sent as a trailer.
   absl::optional<Grpc::Status::GrpcStatus> grpc_status;
   uint64_t grpc_to_http_status = 0;
+  uint64_t put_result_code = response_code;
```
Good idea. Thanks!

cpakulski commented:
/retest

cpakulski commented:
release notes are still needed. I will add them once the PR is ready for merge.

cpakulski commented:
Thanks @paul-r-gall . CI fails now and I am investigating if my changes cause those errors.

cpakulski commented:
/retest

cpakulski commented:
@paul-r-gall . CI passes now. I had to do minor adjustment to a namespace in one of the tests.
